### ISSUE TITLE
Cannot index into a list returned by a function in interpolation expressions

### ISSUE THREAD

**Issue by wr0ngway** [hashicorp/terraform#9850]
Hi there,

Thank you for opening an issue. Please note that we try to keep the Terraform issue tracker reserved for bug reports and feature requests. For general usage questions, please see: https://www.terraform.io/community.html.

### Terraform Version
Terraform v0.7.8

### Affected Resource(s)
- core

### Terraform Configuration Files
```hcl
variable "small_list" {
  type = "list"
  default = [{x = "y"}]
}

variable "filler_list" {
  type = "list"
  default = [{}, {}, {}, {}]
}

output "out2" {
  value = "${concat(var.small_list, var.filler_list)[0]}"
  // also fails
  // value = "${list("1", "2")[0]}"
}
```

### Debug Output
N/A

### Panic Output
N/A

### Expected Behavior
I should be able to get an item from a dynamically generated list

### Actual Behavior

Error loading Terraform: Error loading config: Error loading /Users/mconway/workspace/terraform-test/lists/main.tf: Error reading config for output out2: parse error: syntax error

### Steps to Reproduce
Please list the steps required to reproduce the issue, for example:
1. `terraform apply`

### Important Factoids
Nope, pretty simple example

### References
N/A

**Comment by apparentlymart** [comment]
This is one of the things #11704 aimed to address. Although that wasn't able to land in its current state it is still something I plan to address in future, though likely to be considered as part of a more holistic plan to fix some wrinkles in the design of Terraform's interpolation language.

**Comment by apparentlymart** [comment]
Hi again @wr0ngway!

That "holistic plan" I mentioned has now resulted in a new implementation of the configuration language in master, ready to be included in the forthcoming v0.12.0 release.

I tested the configuration you gave with the v0.12.0-alpha2 prerelease build, and it seems to indeed be working as expected:

```
$ terraform apply

Apply complete! Resources: 0 added, 0 changed, 0 destroyed.

Outputs:

out2 = {}
```

Thanks for reporting this, and sorry for not keeping this issue updated.

**Comment by wr0ngway** [comment]
Cool, thanks!
